ML Zoomcamp Homework 1

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb


In [3]:
pd.__version__

'3.0.5'

In [5]:
df = pd.read_csv('data/car_fuel_efficiency_2026.csv')
df.head()

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0


In [9]:
df.info()
df.isnull().sum()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   model_year           10000 non-null  int64  
 1   origin               10000 non-null  str    
 2   fuel_type            10000 non-null  str    
 3   drivetrain           10000 non-null  str    
 4   num_doors            10000 non-null  int64  
 5   engine_displacement  10000 non-null  int64  
 6   num_cylinders        10000 non-null  int64  
 7   horsepower           9123 non-null   float64
 8   vehicle_weight       10000 non-null  int64  
 9   acceleration         9736 non-null   float64
 10  fuel_efficiency_mpg  10000 non-null  float64
dtypes: float64(3), int64(5), str(3)
memory usage: 859.5 KB


model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

Q2: There are 10.000 records in the dataset

In [20]:
df.nunique()

model_year              50
origin                   3
fuel_type                3
drivetrain               3
num_doors                4
engine_displacement    127
num_cylinders            5
horsepower             153
vehicle_weight         176
acceleration            93
fuel_efficiency_mpg    188
dtype: int64

Q3: There are 3 fuel types presented in the dataset

In [21]:
df.isnull().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

Q4: There are 2 columns in the dataset which have missing values: horse power and acceleration

In [28]:
df.origin.value_counts()

origin
USA       4202
Asia      2994
Europe    2804
Name: count, dtype: int64

In [26]:
# First, we filter the dataset to include only cars from Asia

df_asia = df[df.origin == 'Asia']
df_asia.info()

<class 'pandas.DataFrame'>
Index: 2994 entries, 2 to 9996
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   model_year           2994 non-null   int64  
 1   origin               2994 non-null   str    
 2   fuel_type            2994 non-null   str    
 3   drivetrain           2994 non-null   str    
 4   num_doors            2994 non-null   int64  
 5   engine_displacement  2994 non-null   int64  
 6   num_cylinders        2994 non-null   int64  
 7   horsepower           2729 non-null   float64
 8   vehicle_weight       2994 non-null   int64  
 9   acceleration         2994 non-null   float64
 10  fuel_efficiency_mpg  2994 non-null   float64
dtypes: float64(3), int64(5), str(3)
memory usage: 280.7 KB


In [30]:
# The number of entries in df_asia is correct, according to the previous count of asian cars in the original df.
# Now we can find the max fuel efficiency for cars from Asia.

df_asia['fuel_efficiency_mpg'].describe()

count    2994.000000
mean       30.741015
std         2.835474
min        21.300000
25%        28.700000
50%        30.700000
75%        32.500000
max        41.200000
Name: fuel_efficiency_mpg, dtype: float64

Q5: The max value of fuel effciency in Asian cars is 41.2

In [33]:
# Median value of horsepower
df.horsepower.median()

np.float64(254.0)

In [32]:
# Most frequent value of horsepower
df.horsepower.mode()

0    252.0
Name: horsepower, dtype: float64

In [36]:
n = df.horsepower.mode()[0]
df.horsepower.fillna(n).isnull().sum()

np.int64(0)

In [39]:
df.horsepower.fillna(n).median()

np.float64(252.0)

Q6:
1. Median value of horsepower column (original df): 254.0
2. Most frequent value (mode) of horsepower column: 252.0
4. Median value of horsepower column (modified): 252.0
The median value has decreased after the NaN entries were replaced with the mode value.

In [47]:
df_reduced = df_asia[['model_year', 'vehicle_weight']]
df_reduced.head(7)

,model_year,vehicle_weight
2,1996,4240
6,1992,4310
8,1997,4230
10,1978,4420
15,2007,3950
16,2021,4750
22,2019,4450


In [46]:
X = df_reduced.iloc[0:7,:].to_numpy()
X

array([[1996, 4240],
       [1992, 4310],
       [1997, 4230],
       [1978, 4420],
       [2007, 3950],
       [2021, 4750],
       [2019, 4450]])

In [54]:
XTX = X.T.dot(X)
XTX_inv = np.linalg.inv(XTX)
print(f'XTX:\n{XTX}')
print(f'XTX_inv:\n{XTX_inv}')

XTX:
[[ 28041424  60750580]
 [ 60750580 131950500]]
XTX_inv:
[[ 1.39675281e-05 -6.43071025e-06]
 [-6.43071025e-06  2.96830537e-06]]


In [57]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
w = XTX_inv.dot(X.T).dot(y)
print(f'Sum of weights: {w.sum()}')

Sum of weights: 0.3691969690492609
